labgrid/driver/rawnetworkinterfacedriver.py

```diff
@@ -97,17 +104,26 @@ def record(self, filename, *, count=None, timeout=None):
         Either count or timeout must be specified.
 
         Args:
-            filename (str): name of a file to record to
+            filename (str): name of a file to record to, or None to live stream packets
             count (int): optional, exit after receiving this many number of packets
-            timeout (int): optional, maximum number of seconds to wait for the tcpdump process to
-                           terminate
+            timeout (int): optional, number of seconds to capture packets before tcpdump exits
+        Returns:
+            Popen object of tcpdump process. If filename is None, packets can be read from stdout
         """
         assert count or timeout
 
         try:
-            yield self.start_record(filename, count=count)
+            yield self.start_record(filename, count=count, timeout=timeout)
         finally:
-            self.stop_record(timeout=timeout)
+            # If live streaming packets, there is no reason to wait for tcpdump
+            # to finish, so terminate it as soon as the context is left
+            try:
+                self.stop_record(timeout=0 if filename is None else None)
+            except subprocess.TimeoutExpired:
+                # A timeout is expected if the context is exited early while
+                # piping to stdout
+                if filename is not None:
+                    raise
```

Comment:
I've tested the live streaming feature more: this unfortunately causes exceptions shown in the StepLogger output:

```
[...]
INFO         StepLogger:  → RawNetworkInterfaceDriver.start_record(filename=None count=1 timeout=5) 
INFO         StepLogger:  ← RawNetworkInterfaceDriver.start_record() 
INFO         StepLogger:  → SSHDriver.put(filename='/tmp/pytest-of-bst/pytest-426/test_port_bridge_from_internal1/replay-dut-br0.pcap' remotepath='/tmp/replay-br0.pcap') 
INFO         StepLogger:  ← SSHDriver.put() [0.050s]
INFO         StepLogger:  → SSHDriver.run_check(cmd='tcpreplay --intf1=br0 /tmp/replay-br0.pcap') 
INFO         StepLogger:  ← SSHDriver.run_check() result=['Actual: 1 packets (1523 bytes) sent in 0.000436 seconds', 'Rated: 3493119.2 Bps, 27.94 Mbps, 2293.… [0.169s]
INFO         StepLogger:  → RawNetworkInterfaceDriver.stop_record(timeout=0) 
INFO         StepLogger:  ⚠ RawNetworkInterfaceDriver.stop_record() [0.001s] exception=Command '['ssh', '-x', '-o', 'LogLevel=ERROR', '-o', 'PasswordAuthentication=no', 'myexporter', '--', 'sudo labgrid-raw-interface tcpdump enp2s0f1 1 --timeout 5']' timed out after 0 seconds
[...]
```

This is confusing. To fix this, let's drop the `TimeoutExpired` handling here:

```suggestion
            self.stop_record(timeout=0 if filename is None else None)
```

..and move it to `stop_record()` instead:

```python
@Driver.check_active
@step(args=["timeout"])
def stop_record(self, *, timeout=None):
    """
    Stops previously started tcpdump on bound network interface resource.

    Args:
        timeout (int): optional, maximum number of seconds to wait for the tcpdump process to
                       terminate
    """
    try:
        self._stop(self._record_handle, timeout=timeout)
    except subprocess.TimeoutExpired:
        # If live streaming packets, there is no reason to wait for tcpdump                   
        # to finish, so expect a timeout if piping to stdout
        if self._record_handle.stdout is None:
            raise
    finally:
        self._record_handle = None
```

Reply:
Integrated.